File: day14/14-2.py
initial_positions_and_velocities = '''your input here'''.splitlines()

WIDTH = 101
HEIGHT = 103

robots = []
for robot in initial_positions_and_velocities:
    p, v = robot.split(' ')
    px, py = p[2:].split(',')
    vx, vy = v[2:].split(',')
    robots.append((int(px), int(vx), int(py), int(vy)))

def look_for_tree(simulation_time):
    viz = [['.' for _ in range(WIDTH)] for _ in range(HEIGHT)]
    positions = (((px + vx * simulation_time) % WIDTH, (py + vy * simulation_time) % HEIGHT) for px, vx, py, vy in robots)
    for px, py in positions:
        viz[py][px] = '#'
    
    print(f'__________________________\nTIME: {simulation_time}\n__________________________')
    for row in viz:
        for c in row:
            print(c, end='')
        print()
    
'''
Look for two occurrences:
    a drawing with mostly horizontal lines,
    another with mostly vertical lines.
    
    Example: mostly horizontal at 70, mostly vertical at 10.
'''
for i in range (0, HEIGHT):
    look_for_tree(i)

mostly_horizontal = int(input()) # 70 with my input
'''
The horizontal pattern repeats every HEIGHT iterations,
the vertical pattern repeats every WIDTH iterations.
Find the number where they overlap.
mostly_vertical + i * WIDTH == mostly_horizontal + j * HEIGHT == answer

Example:
(10 + 73 * 101) == (70 + 71 * 103)
== 7383
'''
for i in range(mostly_horizontal, HEIGHT * WIDTH, HEIGHT):
    look_for_tree(i)

File: day14/14-2.c
#include <stdio.h>
#include <string.h>

#define INPUT_SIZE 500
#define WIDTH 101
#define HEIGHT 103
#define FOUR_DOTS 0x2E2E2E2E

struct robot {
    int px;
    int py;
    int vx;
    int vy;
};

void look_for_tree(int simulation_time, struct robot *robots) {
    char viz[HEIGHT][WIDTH + 2]; // +2 for '\n' and '\0'
    memset(viz, FOUR_DOTS, sizeof(viz));
    for (int i = 0; i < INPUT_SIZE; i++)
    {
        struct robot r = robots[i];
        int x = (r.px + r.vx * simulation_time) % WIDTH;
        x += (x < 0) * WIDTH;
        int y = (r.py + r.vy * simulation_time) % HEIGHT;
        y += (y < 0) * HEIGHT;
        viz[y][x] = '#';
    }
    printf("__________________________\nTIME: %d\n__________________________\n", simulation_time);
    for (int i = 0; i < HEIGHT; i++)
    {
        viz[i][WIDTH] = '\n';
        viz[i][WIDTH + 1] = '\0';
        printf("%s", viz[i]);
    }
}

void read_input(struct robot *r);

void main(int argc, char const *argv[])
{
    struct robot robots[INPUT_SIZE];
    read_input(robots);
    // Check the Python file for an explanation on how to find your value
    // using prime number math and visual inspection.
    look_for_tree(7383, robots);
}

void read_input(struct robot *r) {
    FILE *f = fopen("14-input", "r");
    char buffer[32];
    for (int i = 0; i < INPUT_SIZE; i++) {
        fgets(buffer, sizeof(buffer), f);
        sscanf(buffer, "p=%d,%d v=%d,%d", &r[i].px, &r[i].py, &r[i].vx, &r[i].vy);
    }
    fclose(f);
}
